Make defmodel take protocol implementations

plexus opened the pull request:
Pass implementations of protocols directly to `defmodel`, the same way you would
do in `defrecord` or `deftype`. (fixes #9)

Also provides implementations for `empty` (fixes #1) and `apply` (fixes #2).

This updates the documentation, promoting this way of doing things as the
preferred method (vs using extend...merge...IModelDefaults), but still leaves in
an example of how to do things the old-fashioned way.

The defining-models doc has been expanded slightly to further clarify the
relationship between `User` and `(User 1`).

The fill column in that document was irregular but seemed to hover around 120,
so I went with breaking lines at 120 characters. It might be best to ignore
whitespace changes when reviewing this.

Note that as it stands you can implement protocols in `defmodel`, but not
interfaces, since the forms are inserted in `extend` rather than in `defrecord`
itself. To fix this we'll have to separate extensions to IModel from other
interface/protocol implementations.

plexus commented:
Diff ignoring whitespace: https://github.com/metabase/toucan/pull/14/files?w=1

plexus commented:
I'll update VERSION-HISTORY once #13 is merged, or we'll be getting conflicts.

camsaul commented on test/toucan/models_test.clj:
```diff
@@ -159,6 +159,11 @@
+;; Test invoking model as a function with apply
+(expect
+ #toucan.test_models.venue.VenueInstance{:category :store, :name "BevMo", :id 3}
+ (apply Venue [:name "BevMo"]))
```
👍

camsaul commented:
@plexus sorry for the delay. I'll take a look at this and your other PR

camsaul commented on docs/defining-models.md:
```diff
@@ -1,11 +1,10 @@
-Generally, a single Toucan model corresponds to a single table in your database. Common models might
-be things like `User` or `Venue`.
+Generally, a single Toucan model corresponds to a single table in your database. Common models might be things like
+`User` or `Venue`.
```
what's the reason behind changing these lines? 

it makes it a lot harder to review PRs when things are randomly changed like this for no reason

camsaul commented on docs/defining-models.md:
````diff
@@ -20,49 +19,81 @@ The `defmodel` macro is used to define a new model:
 In other words, both of the following work:
 
-```clojure
+``` clojure
````
why

camsaul commented on docs/defining-models.md:
```diff
@@ -20,49 +19,81 @@ The `defmodel` macro is used to define a new model:
-Refer to the documentation below for an overview of the various methods provided by `IModel` and how to use them to customize
-the behavior of your models.
+Refer to the documentation below for an overview of the various methods provided by `IModel` and how to use them to
+customize the behavior of your models.
```
again, random edits like this that don't affect the way the markdown renders make this PR a lot bigger (and thus a lot harder to review), so what's the reason behind them?

camsaul commented on docs/defining-models.md:
```diff
@@ -292,13 +320,14 @@ any desired changes.
-(extend (class User)
-  IModel (merge IModelDefaults
-                {:pre-insert pre-insert}))
+(defn- )
```
did you mean to remove this?

camsaul commented on docs/defining-models.md:
```diff
@@ -292,13 +320,14 @@ any desired changes.
+  (pre-insert [user]
+    (let [defaults {:version 1}]
+      (merge defaults user) ; set some default values"
+      )))
```
hanging parens = bad

this should look like `)))) ; set some default values`

camsaul commented on src/toucan/models.clj:
```diff
@@ -371,6 +373,45 @@
 ;;; ========================================================================================================================
 
+(defn- ifn-invoke-forms
+  "Macro helper, generates
```
generates what?

camsaul commented on src/toucan/models.clj:
```diff
@@ -371,6 +373,45 @@
+        (invoke-model-or-instance this# id#))
+       (~'invoke [this# arg1# arg2#]
+        (invoke-model-or-instance this# arg1# arg2#))
+       ,,,"
```
I think you meant to put `...`

plexus replied:
`,,,` is called a "fogus comma", it's some used in Clojure docs (e.g. [this one](https://clojure.org/guides/threading_macros), as well by [PragProg](https://media.pragprog.com/titles/vmclojeco/intro.pdf)

![2017-06-01-145244_615x93_scrot](https://cloud.githubusercontent.com/assets/32212/26680510/00fef6d2-46da-11e7-846e-ad0598f6f825.png)

I can change it though if you don't like it.

camsaul replied:
Got it. I haven't seen that before, but it works for me 👍

camsaul commented:
@plexus don't worry about updating the VERSION-HISTORY, I'll do it when I push out a new release

plexus commented:
I cleaned things up that were pointed out in the review. My apologies for making seemingly arbitrary changes, I reverted them. I tend to use emacs' fill-paragraph functionality a lot when writing docs. However in this case it seems the docs are wrapped by hand, so their fill-column isn't consistent.

camsaul commented:
@plexus Sorry it took me so long to merge this things have been a little hectic around here. But it looks good. Thanks!
